client/wfm/src/service/WfmService.ts

```diff
@@ -0,0 +1,248 @@
+    return this.workorderService.read(workorderId)
+      .then(workorder => Promise.all([
+        this.workflowService.read(workorder.workflowId),
+        this.resultService.readByWorkorder(workorderId)
+      ]).then(([workflow, result]) => ({
+        workflow,
+        workorder,
+        result,
+        nextStepIndex: this.executeStepReview(workflow.steps, result).nextStepIndex
+      })));
+  }
+
+  /**
+   *
+   * Going to the previous step of a workorder.
+   *
+   * @param workorderId - The ID of the workorder to switch to the previous step for
+   */
+  public previousStep(workorderId): Promise<StepOperationData> {
+    const self = this;
+    return this.workorderSummary(workorderId).then(summary => {
+      const { workorder, workflow, result } = summary;
+      let { nextStepIndex } = summary;
+
+      if (!result) {
+        // No result exists, The workflow should have been started
+        return Promise.reject(new Error('No result exists for workflow ' + workorderId +
+          '. The workflow back topic can only be used for a workflow that has begun'));
+      }
+
+      // -1 is a special value for 'no next step'
+      nextStepIndex = _.max([nextStepIndex - 1, -1]) || -1; // _.max returns `number?`
+
+      return self.resultService.update(result)
+        .then(() => ({
+          workorder,
+          workflow,
+          result,
+          nextStepIndex,
+          step: nextStepIndex > -1 ? workflow.steps[nextStepIndex] : null
+        }));
+    });
+  }
+
+  /**
+   * Completing a single step for a workorder.
+   */
+  public completeStep(parameters: CompleteStepParams): Promise<StepOperationData> {
+    const workorderId = parameters.workorderId;
+    const stepCode = parameters.stepCode;
+    const submission = parameters.submission;
+    return Promise.all([
+      this.userService.readUser().then(profileData => profileData.id),
+      this.workorderSummary(workorderId)
+    ]).then(([userId, summary]) => {
+      const { workorder, workflow, result, nextStepIndex } = summary;
+
+      if (!result) {
+        // No result exists, The workflow should have been started
+        return Promise.reject(new Error('No result exists for workorder ' + workorderId +
+          '. The workflow done topic can only be used for a workflow that has begun'));
+      }
+
+      const step = _.find(workflow.steps, s => s.code === stepCode);
+
+      // If there is no step, then this step submission is invalid.
+      if (!step) {
+        // No result exists, The workflow should have been started
+        return Promise.reject(new Error('Invalid step to assign completed data for workorder ' + workorderId +
+          ' and step code ' + stepCode));
+      }
+
+      // Got the workflow, now we can create the step result.
+      const stepResult = {
+        step,
+        submission,
+        status: STATUS.COMPLETE,
+        timestamp: new Date().getTime(),
+        submitter: userId
```

Comment:
`submitter` imho is sounds strange. If that's not a big task we may change it to `completedBy`, `owner`, `userReference` etc.